### ISSUE TITLE
False positive CastToNullableType on safe typealias cast

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#6676]
## Expected Behavior
No report, it's a valid and equivalent cast.

## Observed Behavior
```
Use the safe cast ('as? Alias') instead of 'as Alias?'. [CastToNullableType]
```
The cast is very safe, it's so safe in fact, that Kotlin says
```
> Task :compileKotlin
w: ... No cast needed
```
This is why the repro suppresses.

## Steps to Reproduce
```kotlin
typealias Alias = String

fun f(s: String?) {
	@Suppress("USELESS_CAST") // Casts is useful to convert String to typealias.
	val a = s as Alias?
	val l /*List<Alias?>*/ = listOf(a)
	println(l)
}
```
The reason for the cast is to get the hint in the comment. This is a trivial example, the benefit is not easily visible, see below for more.

## Context
I have some complex logic chain where there's a lot of type changes, because of map/filter logic.
I also use some external types which contain string properties representing different concepts.
I created some private aliases to help understand the code and I'm casting where necessary to see the alias types in the type hints.
![image](https://github.com/detekt/detekt/assets/2906988/59689856-67fc-41fc-8567-a45b21a7e082)

## Your Environment

* Version of detekt used: 1.23.4
* Version of Gradle used (if applicable): 8.5
* Gradle scan link (add `--scan` option when running the gradle task): N/A
* Operating System and version: any
* Link to your project (if it's a public repository): N/A, full repro above

**Comment by atulgpt** [comment]
@TWiStErRob fixing this would require converting this rule to `@RequiresTypeResolution`

Also, I think the below failing TCs should be fixed

```kotlin
// https://github.com/detekt/detekt/issues/6676
@Test
fun `cast to same type in alias form allowed`() {
    val code = """
        typealias Alias = String

        fun test(s: String?) {
            @Suppress("USELESS_CAST") // Casts is useful to convert String to typealias.
            val a = s as Alias?
            print(a)
        }
        """.trimIndent()
    val findings = subject.compileAndLint(code)
    assertThat(findings).isEmpty()
}

@Test
fun `cast to parent type in alias form allowed`() {
    val code = """
        typealias Alias = CharSequence

        fun test(s: String?) {
            val a = s as Alias?
            print(a)
        }
        """.trimIndent()
    val findings = subject.compileAndLint(code)
    assertThat(findings).isEmpty()
}

@Test
fun `cast to different type in alias form not allowed`() {
    val code = """
        typealias Alias = String

        fun test(s: Any?) {
            val a = s as Alias?
            print(a)
        }
        """.trimIndent()
    val findings = subject.compileAndLint(code)
    assertThat(findings).hasSize(1)
}
```

**Comment by TWiStErRob** [comment]
Re test, agreed.

Re type-resolution: I think that's a fair change. Anything dealing with nullability needs to be aware of the type system.
